### PR TITLE
Compute floor_divide in float for Half and BFloat16

### LINKED ISSUE #193754: floor_divide on float16/bfloat16 depends on tensor length, and the vectorized result can exceed the quotient
### 🐛 Describe the bug

`torch.floor_divide` on float16 and bfloat16 gives a different answer depending on how long the tensor is, and the vectorized answer can be larger than the quotient it is supposed to be the floor of.

```python
import torch

a = torch.full((2,), 560.0, dtype=torch.bfloat16)
b = torch.full((2,), 3.0, dtype=torch.bfloat16)
print(torch.floor_divide(a, b)[0].item())        # 187.0
print(torch.floor_divide(a[:1], b[:1])[0].item())  # 186.0
```

560 and 3 are both exact in bfloat16, and 560 / 3 is 186.67, so the floor is 186. The length-2 result 187 gives `187 * 3 = 561`, which is greater than 560. The reconstruction breaks with it:

```python
q = torch.floor_divide(a, b)
r = torch.remainder(a, b)
print((q * b + r)[0].item())  # 563.0, not 560.0
```

float16 does the same:

```python
a = torch.full((2,), 2158.0, dtype=torch.float16)
b = torch.full((2,), 3.0, dtype=torch.float16)
print(torch.floor_divide(a, b)[0].item())        # 718.0
print(torch.floor_divide(a[:1], b[:1])[0].item())  # 719.0   correct
```

On 4000 random pairs per dtype, comparing a length-1 tensor against a length-64 tensor of the same value:

| dtype | pairs that disagree |
|---|---|
| float16 | 39 / 4000 |
| bfloat16 | 65 / 4000 |
| float32 | 0 / 4000 |
| float64 | 0 / 4000 |

float32 and float64 being clean is what points at the cause. The scalar path calls `c10::div_floor_floating(a, b)` with `scalar_t` = `Half` or `BFloat16`, and `std::fmod` on those promotes to float, so the whole computation runs in fp32 and only the return value narrows:

https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L341

The vectorized path runs in `Vectorized<Half>` / `Vectorized<BFloat16>`, so `a.fmod(b)` and `(a - mod) / b` each round back to 8 or 11 mantissa bits:

https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L344

This class is already recognized in the same file. The SVE bf16 specialization does exactly the needed widening, for the reason given in its comment:

https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L277-L295

It is guarded by `CPU_CAPABILITY_SVE256 && __ARM_FEATURE_BF16`, so every other build gets the narrow computation, and float16 gets it everywhere. Making that specialization unconditional and adding one for `Half` would line the vec path up with the scalar path.

I know results are not guaranteed bitwise identical across vectorization. This is not a last-bit difference. The result is off by a whole unit and is not a floor of anything.

Found by a differential tester that varies tensor length while holding the values fixed. Related to #193753, which is a different root cause in the same file.

### Versions

torch 2.13.0, CPU, macOS arm64.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### PR DESCRIPTION
**Pull request by krrishapatel**
Fixes #193754.

`torch.floor_divide` on `bfloat16` and `float16` gives a different answer depending on the length of the input:

```python
>>> import torch
>>> a, b = torch.full((2,), 560.0, dtype=torch.bfloat16), torch.full((2,), 3.0, dtype=torch.bfloat16)
>>> torch.floor_divide(a, b)[0]
tensor(187., dtype=torch.bfloat16)
>>> torch.floor_divide(a[:1], b[:1])[0]
tensor(186., dtype=torch.bfloat16)
```

186 is right.

### Cause

`div_floor_kernel` has two paths. When the divisor is a scalar it dispatches through `AT_DISPATCH_REDUCED_FLOATING_TYPES` and computes in `opmath_t`, so `bfloat16` runs in float. The tensor-tensor path computes in `scalar_t`.

A size-1 tensor is squeezed to 0-dim, which makes `iter.is_scalar(2)` true, so it takes the widened path. That is the whole length dependence.

Staying in `bfloat16` is what gets the wrong answer. From NOTE: [Floor Division in Python], the quotient is `(a - fmod(a, b)) / b`. For 560 and 3 that is `558 / 3`, but 558 needs 10 significand bits and `bfloat16` has 8, so it rounds to 560, and `560 / 3` rounds to 187.

`std::fmod` does not save this. `c10::BFloat16` has its own overload returning `c10::BFloat16`, so every step in `c10::div_floor_floating` rounds.

The error is not bounded to one unit either. Over 4096 random values the largest difference from float math was 64.

### Fix

Widen both loops in the tensor-tensor path to `opmath_t`, matching the scalar divisor path. Same one-line cast on CUDA, whose general path had the same split against its own `is_cpu_scalar(2)` branch.

In the vectorized loop this replaces the `CPU_CAPABILITY_SVE256 && __ARM_FEATURE_BF16` specialization of `div_floor_floating_vec`, which already did exactly this widening for one backend and one dtype. Since it did not widen the scalar loop next to it, SVE256 has the same length dependence today, just in the other direction.

Written as `if constexpr` on `is_reduced_floating_point_v` rather than two specializations, so the float body is not instantiated for `Half` and `BFloat16` at all. `git diff -w` on that hunk is small, the body is only reindented.

float and double are unaffected. `opmath_type<float>` is `float`, so the cast is a no-op and the `if constexpr` picks the existing body.

### Correctness

4000 random pairs per dtype, comparing the tensor-tensor result against the same op elementwise on 0-dim tensors, and against float math narrowed once:

| | vs 0-dim | vs float |
|---|---|---|
| bfloat16 before | 264 / 4000 | 264 / 4000 |
| bfloat16 after | 0 | 0 |
| float16 before | 40 / 4000 | 40 / 4000 |
| float16 after | 0 | 0 |
| float32, float64 | 0 both | 0 both |

`560 // 3` in `bfloat16` is now 186 at lengths 1, 2, 3, 8, 16, 64 and 1000, and at 0-dim.

### Performance

This is faster, not slower. Single threaded, arm64, best of 7.

| n = 2^20 | before | after | |
|---|---|---|---|
| bfloat16 | 25262 us | 2779 us | 9.1x |
| float16 | 21006 us | 2244 us | 9.4x |
| float32 | 2145 us | 2100 us | unchanged |
| float64 | 4701 us | 4552 us | unchanged |

At n = 4096: bfloat16 51.5 -> 11.8 us, float16 31.6 -> 9.1 us.

`Vectorized<BFloat16>::fmod` is `map2(q, std::fmod)` and `::floor` is `map(floor_impl)`, so the old path spilled each vector to memory and called scalar libm per lane, and `std::fmod` on `BFloat16` converted to float and back anyway. Converting once up front gets `Sleef_fmod` and `vrndmq_f32` on real float vectors. This is what the SVE comment predicted, that most of these instructions end up as f32 anyway.

### Tests

`test_div_floor_reduced_precision` in `test/test_binary_ufuncs.py`. It compares against float math narrowed once with `atol=0, rtol=0`, since the default `bfloat16` tolerance is wide enough to hide an off-by-one quotient. It also checks lengths 1, 2 and 3 so the scalar loop and the vectorized loop are both covered.

Before this change it fails on 240 / 4096 elements for `bfloat16` and 18 / 4096 for `float16`.

`test_binary_ufuncs.py` passes, 12713 passed. `test_ops.py -k "floor_divide or div_floor_rounding"` passes, 169 passed.

I do not have a CUDA machine, so that one line is only reasoned about and left to CI.

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!
